src/math_utils.rs

```diff
@@ -182,6 +185,176 @@ pub fn sqrt(n: &Felt252) -> Felt252 {
     }
 }
 
+// Adapted from sympy _sqrt_prime_power with k == 1
+pub fn sqrt_prime_power(a: &BigUint, p: &BigUint) -> Option<BigUint> {
+    let two = BigUint::from(2_u32);
+    let a = a.mod_floor(p);
+    if p == &two {
+        return Some(a);
+    }
+    if !(a < two || (a.modpow(&(p - 1_u32).div_floor(&two), p)).is_one()) {
+        return None;
+    };
+
+    if p.mod_floor(&BigUint::from(4_u32)) == 3_u32.into() {
+        let res = a.modpow(&(p + 1_u32).div_floor(&BigUint::from(4_u32)), p);
+        return Some(min(res.clone(), p - res));
+    };
+
+    if p.mod_floor(&BigUint::from(8_u32)) == 5_u32.into() {
+        let sign = a.modpow(&(p - 1_u32).div_floor(&BigUint::from(4_u32)), p);
+        if sign.is_one() {
+            let res = a.modpow(&(p + 3_u32).div_floor(&BigUint::from(8_u32)), p);
+            return Some(min(res.clone(), p - res));
+        } else {
+            let b = (4_u32 * &a).modpow(&(p - 5_u32).div_floor(&BigUint::from(8_u32)), p);
+            let x = (2_u32 * &a * b).mod_floor(p);
+            if x.modpow(&two, p) == a {
+                return Some(x);
+            }
+        }
+    };
+
+    Some(sqrt_tonelli_shanks(&a, p))
+}
+
+fn sqrt_tonelli_shanks(n: &BigUint, prime: &BigUint) -> BigUint {
+    // Based on Tonelli-Shanks' algorithm for finding square roots
+    // and sympy's library implementation of said algorithm.
+    if n.is_zero() || n.is_one() {
+        return n.clone();
+    }
+    let s = trailing(prime - 1_u32);
```

Comment:
```suggestion
    let s = (prime - 1_u32).trailing_zeros();
```
https://docs.rs/num-bigint/0.4.3/num_bigint/struct.BigUint.html#method.trailing_zeros

Reply:
Changed!